### PR TITLE
Upgrade to arrow 15

### LINKED ISSUE #2453: Incorrect Parquet Projection For Nested Types
**Describe the bug**

```
use datafusion::prelude::{ParquetReadOptions, SessionContext};

#[tokio::test]
async fn temp() {
    let ctx = SessionContext::new();

    ctx.register_parquet("patient", "part-00000-f6337bce-7fcd-4021-9f9d-040413ea83f8-c000.snappy.parquet",
                         ParquetReadOptions::default()).await.unwrap();

    let df = ctx.sql("SELECT patient.meta FROM patient LIMIT 10").await.unwrap();
    df.show().await.unwrap();
}
```

Where part-00000-f6337bce-7fcd-4021-9f9d-040413ea83f8-c000.snappy.parquet is the [parquet file](https://github.com/apache/arrow-datafusion/files/8626500/part-00000-f6337bce-7fcd-4021-9f9d-040413ea83f8-c000.snappy.parquet.zip) provided by @kesavkolla in https://github.com/apache/arrow-datafusion/issues/2439

This fails with

```
called `Result::unwrap()` on an `Err` value: ArrowError(ExternalError(ArrowError(InvalidArgumentError("column types must match schema types, expected Struct([Field { name: \"id\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"extension\", data_type: List(Field { name: \"element\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"versionId\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"lastUpdated\", data_type: Timestamp(Nanosecond, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"source\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"profile\", data_type: List(Field { name: \"element\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"security\", data_type: List(Field { name: \"element\", data_type: Struct([Field { name: \"id\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"extension\", data_type: List(Field { name: \"element\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"system\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"version\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"code\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"display\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"userSelected\", data_type: Boolean, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }]), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"tag\", data_type: List(Field { name: \"element\", data_type: Struct([Field { name: \"id\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"extension\", data_type: List(Field { name: \"element\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"system\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"version\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"code\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"display\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"userSelected\", data_type: Boolean, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }]), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }]) but found Struct([Field { name: \"id\", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }]) at column index 0"))))
```

The problem arises because ParquetExec is passing the projection indices for the arrow schema to get_record_reader_by_columns which instead expects parquet column indexes. In the presence of nested types, these are not the same thing.

This is further complicated by https://github.com/apache/arrow-rs/issues/1652 and https://github.com/apache/arrow-rs/issues/1651

**To Reproduce**

Run the code above

**Expected behavior**

The code should not error

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2530

Closes #2453

# Rationale for this change

Arrow 15 contains breaking changes, lets validate them prior to release.

# What changes are included in this PR?

Updates to the current arrow-rs master

# Are there any user-facing changes?

No

# Does this PR break compatibility with Ballista?

Possibly, I think I need to modify ballista to use the re-exports from Datafusion

### CODE REVIEW

File: datafusion/core/src/physical_plan/common.rs [hunk 1]
```diff
@@ -365,7 +365,7 @@ mod tests {
         let expected = Statistics {
             is_exact: true,
             num_rows: Some(3),
-            total_byte_size: Some(416), // this might change a bit if the way we compute the size changes
+            total_byte_size: Some(464), // this might change a bit if the way we compute the size changes
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/common.rs:368]: The changes to DataType have made it bigger, I don't think this is a problem

**Comment by alamb** [comment]
FWIW arrow 15 is now released https://github.com/apache/arrow-rs/issues/1727

**Comment by tustvold** [comment]
Currently blocked on https://github.com/apache/arrow-ballista/pull/48

**Review by Jimexist — APPROVED** [review]

**Review by alamb — APPROVED** [review]
Thank you @tustvold

**Pull request merged**